dll/win32/shlwapi/propbag.cpp

```diff
@@ -580,3 +580,281 @@ SHCreatePropertyBagOnRegKey(
 
     return pRegBag->QueryInterface(riid, ppvObj);
 }
+
+/**************************************************************************
+ *  SHGetIniStringUTF7W (SHLWAPI.473)
+ */
+EXTERN_C DWORD WINAPI
+SHGetIniStringUTF7W(
+    _In_opt_z_ LPCWSTR lpAppName,
+    _In_z_ LPCWSTR lpKeyName,
+    _Out_ LPWSTR lpReturnedString,
+    _In_ DWORD nSize,
+    _In_z_ LPCWSTR lpFileName)
+{
+    if (*lpKeyName == L'@') // UTF-7
+        return SHGetIniStringW(lpAppName, lpKeyName + 1, lpReturnedString, nSize, lpFileName);
+
+    return GetPrivateProfileStringW(lpAppName, lpKeyName, L"", lpReturnedString, nSize, lpFileName);
+}
+
+/**************************************************************************
+ *  SHSetIniStringUTF7W (SHLWAPI.474)
+ */
+EXTERN_C BOOL WINAPI
+SHSetIniStringUTF7W(
+    _In_z_ LPCWSTR lpAppName,
+    _In_z_ LPCWSTR lpKeyName,
+    _In_opt_z_ LPCWSTR lpString,
+    _In_z_ LPCWSTR lpFileName)
+{
+    if (*lpKeyName == L'@') // UTF-7
+        return SHSetIniStringW(lpAppName, lpKeyName + 1, lpString, lpFileName);
+
+    return WritePrivateProfileStringW(lpAppName, lpKeyName, lpString, lpFileName);
+}
+
+class CIniPropertyBag : public CBasePropertyBag
+{
+protected:
+    LPWSTR m_pszFileName;
+    LPWSTR m_pszSection;
+    BOOL m_bAlternateStream;
+
+    static BOOL LooksLikeAnAlternateStream(LPCWSTR pszStart)
+    {
+        LPCWSTR pch = StrRChrW(pszStart, NULL, L'\\');
+        if (!pch)
+            pch = pszStart;
+        return StrChrW(pch, L':') != NULL;
+    }
+
+    HRESULT
+    _GetSectionAndName(
+        LPCWSTR pszStart,
+        LPWSTR pszSection,
+        UINT cchSectionMax,
+        LPWSTR pszName,
+        UINT cchNameMax);
+
+public:
+    CIniPropertyBag(DWORD dwMode)
+        : CBasePropertyBag(dwMode)
+        , m_pszFileName(NULL)
+        , m_pszSection(NULL)
+        , m_bAlternateStream(FALSE)
+    {
+    }
+
+    ~CIniPropertyBag() override
+    {
+        ::LocalFree(m_pszFileName);
+        ::LocalFree(m_pszSection);
+    }
+
+    HRESULT Init(LPCWSTR pszIniFile, LPCWSTR pszSection);
+
+    STDMETHODIMP Read(
+        _In_z_ LPCWSTR pszPropName,
+        _Inout_ VARIANT *pvari,
+        _Inout_opt_ IErrorLog *pErrorLog) override;
+
+    STDMETHODIMP Write(_In_z_ LPCWSTR pszPropName, _In_ VARIANT *pvari) override;
+};
+
+HRESULT CIniPropertyBag::Init(LPCWSTR pszIniFile, LPCWSTR pszSection)
+{
+    m_pszFileName = StrDupW(pszIniFile);
+    if (!m_pszFileName)
+        return E_OUTOFMEMORY;
+
+    m_bAlternateStream = LooksLikeAnAlternateStream(m_pszFileName);
```

Comment:
This needs an explanation

Reply:
ADS: Alternate Data Stream
https://www.sans.org/blog/alternate-data-streams-overview/

Reply:
A colon and some trailing characters after a pathname has a special meaning in NTFS filesystem.

For example, 
`C:\Windows\SafeFile.txt:EvilFile.txt` is an invisible file from a normal user.